boards/posix/native_posix/hw_models_top.c

```diff
@@ -92,12 +98,17 @@ static void hwm_sleep_until_next_timer(void)
 		/* LCOV_EXCL_STOP */
 	}
 
-	if (signaled_end || (simu_time > end_of_time)) {
+	_SAFE_CALL(pthread_mutex_lock(&exit_mutex));
+	if (signaled_end || (simu_time > end_of_time) || hwm_exit_request) {
+		int exit_code = hwm_exit_request ? hwm_exit_code : 0;
+
+		_SAFE_CALL(pthread_mutex_unlock(&exit_mutex));
 		posix_print_trace("\nStopped right after %.3Lfs\n",
 				((long double)end_of_time)/1.0e6);
 
-		posix_exit(0);
+		posix_exit(exit_code);
 	}
+	_SAFE_CALL(pthread_mutex_unlock(&exit_mutex));
```

Comment:
This one is not needed if you only want to protect the hwm_exit code, and just have the mutex_lock call inside the if.